packages/backend-core/src/utils/utils.ts

```diff
@@ -237,3 +237,17 @@ export function timeout(timeMs: number) {
 export function isAudited(event: Event) {
   return !!AuditedEventFriendlyName[event]
 }
+
+export function hasCircularStructure(json: any) {
+  if (typeof json !== "object") {
+    return false
+  }
+  try {
+    JSON.stringify(json)
+  } catch (err) {
+    if (err instanceof Error && err?.message.includes("circular structure")) {
```

Comment:
This makes me nervous. I was hoping `lodash` would have something that could do this in a more concrete way but they appear to recommend this exact approach instead: https://github.com/lodash/lodash/issues/1548.

Given that there's a test and I imagine the regressive behaviour would be an infinite loop, we'd catch this as soon as it became a problem. Still, gives me the shivers.

Reply:
I think its pretty safe as `JSON.stringify` has always been protected against this - its impossible for it to generate a string when it contains a circular structure, and IMO keeping track of the state isn't always that easy, I prefer this method for managing rows than just trusting Koa to look after it correctly.

Reply:
I'm more worried about the error message changing and we start to incorrectly let objects through.